cert/cert.go

```diff
@@ -19,44 +37,65 @@ import (
 //   - User public key is mapped to the RawSubjectPublicKeyInfo field
 //   - Raw PK token is mapped to the SubjectKeyId field
 func CreateX509Cert(pkToken *pktoken.PKToken, signer crypto.Signer) ([]byte, error) {
-	pkTokenJSON, err := json.Marshal(pkToken)
+	template, err := PktToX509Template(pkToken)
+
+	// create a self-signed X.509 certificate
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, signer.Public(), signer)
+	if err != nil {
+		return nil, fmt.Errorf("error creating X.509 certificate: %w", err)
+	}
+	certBlock := &pem.Block{Type: "CERTIFICATE", Bytes: certDER}
+	return pem.EncodeToMemory(certBlock), nil
+}
+
+// PktToX509Template takes a PK Token and returns a X.509 certificate template
+// with the fields of the template set to the values in the X509
+func PktToX509Template(pkt *pktoken.PKToken) (*x509.Certificate, error) {
+	pktJson, err := json.Marshal(pkt)
 	if err != nil {
 		return nil, fmt.Errorf("error marshalling PK token to JSON: %w", err)
 	}
 
-	// get subject identitifer from pk token
-	var payload struct {
-		Subject string `json:"sub"`
+	// get subject identifier from pk token
+	idtClaims := new(client.OidcClaims)
+	if err := json.Unmarshal(pkt.Payload, idtClaims); err != nil {
+		return nil, err
+	}
+
+	cic, err := pkt.GetCicValues()
+	if err != nil {
+		return nil, err
 	}
-	if err := json.Unmarshal(pkToken.Payload, &payload); err != nil {
+	upk := cic.PublicKey()
+	var rawkey interface{} // This is the raw key, like *rsa.PrivateKey or *ecdsa.PrivateKey
+	if err := upk.Raw(&rawkey); err != nil {
 		return nil, err
 	}
 
 	// encode ephemeral public key
-	ecPub, err := x509.MarshalPKIXPublicKey(signer.Public())
+	ecPub, err := x509.MarshalPKIXPublicKey(rawkey)
 	if err != nil {
 		return nil, fmt.Errorf("error marshalling public key: %w", err)
 	}
-
-	template := x509.Certificate{
+	template := &x509.Certificate{
 		SerialNumber:            big.NewInt(1),
-		Subject:                 pkix.Name{CommonName: payload.Subject},
+		Subject:                 pkix.Name{CommonName: idtClaims.Subject},
 		RawSubjectPublicKeyInfo: ecPub,
 		NotBefore:               time.Now(),
 		NotAfter:                time.Now().Add(365 * 24 * time.Hour), // valid for 1 year
 		KeyUsage:                x509.KeyUsageDigitalSignature,
 		ExtKeyUsage:             []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning},
 		BasicConstraintsValid:   true,
-		DNSNames:                []string{payload.Subject},
+		DNSNames:                []string{idtClaims.Subject},
 		IsCA:                    false,
-		SubjectKeyId:            pkTokenJSON,
+		ExtraExtensions: []pkix.Extension{{
```

Comment:
I should probably make this optional and have it controllable via a bool arg.

Reply:
This is probably OK without making it optional since usage here will always be in the context of a PK Token that has been issued by a provider.